Ordinal arrays that use ints greater than 16 bit error.

Using 32/64 bit ints in the size of an array cause a compiler error. Having both changes the error to an assertion defect.

### Example
```nim
var e: array[int32, byte]
var f: array[uint32, byte]
var g: array[int64, byte]
var h: array[uint64, byte]
```

### Current Output
```
# With 64bits
Error: unhandled exception: int128.nim(72, 11) `arg.sdata(2) == 0` out of range [AssertionDefect]

# With just 32bits
/usercode/nimcache/@min.nim.c.o: In function `main':
@min.nim.c:(.text+0xed): relocation truncated to fit: R_X86_64_PC32 against symbol `cmdLine' defined in COMMON section in /usercode/nimcache/@min.nim.c.o
@min.nim.c:(.text+0xf6): relocation truncated to fit: R_X86_64_PC32 against symbol `cmdCount' defined in COMMON section in /usercode/nimcache/@min.nim.c.o
@min.nim.c:(.text+0x101): relocation truncated to fit: R_X86_64_PC32 against symbol `gEnv' defined in COMMON section in /usercode/nimcache/@min.nim.c.o
collect2: error: ld returned 1 exit status
Error: execution of an external program failed: 'gcc   -o /usercode/in  /usercode/nimcache/stdlib_system.nim.c.o /usercode/nimcache/@min.nim.c.o    -ldl'

```
Tried devel and stable.

--- a/compiler/semtypes.nim
+++ b/compiler/semtypes.nim
@@ -415,13 +415,16 @@ proc semArray(c: PContext, n: PNode, prev: PType): PType =
     if indxB.kind in {tyGenericInst, tyAlias, tySink}: indxB = skipModifier(indxB)
     if indxB.kind notin {tyGenericParam, tyStatic, tyFromExpr} and
         tfUnresolved notin indxB.flags:
-      if indxB.skipTypes({tyRange}).kind in {tyUInt, tyUInt64}:
-        discard
-      elif not isOrdinalType(indxB):
+      if not isOrdinalType(indxB):
         localError(c.config, n[1].info, errOrdinalTypeExpected % typeToString(indxB, preferDesc))
       elif enumHasHoles(indxB):
         localError(c.config, n[1].info, "enum '$1' has holes" %
                    typeToString(indxB.skipTypes({tyRange})))
+      elif indxB.kind != tyRange and
+          lengthOrd(c.config, indxB) > high(uint16).int:
+        # assume range type is intentional
+        localError(c.config, n[1].info,
+          "index type '$1' for array is too large" % typeToString(indxB))
     base = semTypeNode(c, n[2], nil)
     # ensure we only construct a tyArray when there was no error (bug #3048):
     # bug #6682: Do not propagate initialization requirements etc for the

--- a/tests/array/tlargeindex.nim
+++ b/tests/array/tlargeindex.nim
@@ -1,0 +1,18 @@
+discard """
+  cmd: "nim check --hints:off $file"
+"""
+
+# issue #17163
+var e: array[int32, byte] #[tt.Error
+             ^ index type 'int32' for array is too large]#
+var f: array[uint32, byte] #[tt.Error
+             ^ index type 'uint32' for array is too large]#
+var g: array[int64, byte] #[tt.Error
+             ^ index type 'int64' for array is too large]#
+var h: array[uint64, byte] #[tt.Error
+             ^ index type 'uint64' for array is too large]#
+
+# crash in issue #23204
+proc y[N](): array[N, int] = default(array[N, int]) #[tt.Error
+                                           ^ index type 'int' for array is too large]#
+discard y[int]()

